stability: shush senseless log chatter

spencerkimball opened the pull request:
Changed Raft transport to avoid busy logging when node addresses aren't
available via gossip. It now retries with an exponential backoff and
only logs during availability state changes.

Similarly removed endless repetition of gossip logging in the event
that bootstrap hosts don't resolve (happens when you migrate data dirs).

Cleaned up a TODO to purge old, invalid hosts from the bootstrap info
host addresses.

No more repeated warnings on stalled gossip. We now just log the
state transitions.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8021)

<!-- Reviewable:end -->

spencerkimball committed:
stability: quiet overly busy log chatter

Changed Raft transport to avoid busy logging when node addresses aren't
available via gossip. It now uses a circuit breaker to retry with an
exponential backoff and only logs during availability state changes.

Similarly removed endless repetition of gossip logging in the event
that bootstrap hosts don't resolve (happens when you migrate data dirs).

Cleaned up a TODO to purge old, invalid hosts from the bootstrap info
host addresses.

No more repeated warnings on stalled gossip. We now just log the
state transitions.

tamird commented:
Reviewed 4 of 8 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[gossip/gossip.go, line 460 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblVWY_L40McLHqImQ:-KNblVWY_L40McLHqImR:1747057617) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/gossip.go#L460)):*
> ```Go
> 				newAddrs = append(newAddrs, addr)
> 			} else {
> 				log.Infof(context.TODO(), "purging invalid bootstrap address %s", addr)
> ```

throughout: we shouldn't be introducing new `context.TODO()` wherever we're logging anything. can a context be plumbed here?

---

*[gossip/storage_test.go, line 230 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbjdJK-q4hVxtJ2Wjx:-KNbjdJLxMqMOtbpguLG:339957993) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/storage_test.go#L230)):*
> ```Go
> 		if err := stores[i].WriteBootstrapInfo(&gossip.BootstrapInfo{
> 			Addresses: []util.UnresolvedAddr{
> 				util.MakeUnresolvedAddr("tcp", network.Nodes[(i+1)%3].Addr.String()), // node i+1 address
> ```

why `% 3`? this `3` is `len(network.Nodes)`, right? avoid magic numbers.

---

*[gossip/storage_test.go, line 250 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbjTS5hf3XGtoi7UIH:-KNbjTS5hf3XGtoi7UII:1376113905) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/storage_test.go#L250)):*
> ```Go
> 	util.SucceedsSoon(t, func() error {
> 		for i := range stores {
> 			p := &stores[i]
> ```

why not `for i, p := range stores`?

---

*[gossip/storage_test.go, line 257 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbjHbN4aaFM-LKF4BT:-KNbjHbN4aaFM-LKF4BU:106453308) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/storage_test.go#L257)):*
> ```Go
> 			for _, addr := range p.Info().Addresses {
> 				if addr.String() == invalidAddr {
> 					log.Infof("node %d has invalid address", i)
> ```

remove (this also doesn't compile)

---

*[gossip/simulation/network.go, line 183 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbkG6BaYxe4kcEi8R7:-KNbkG6BaYxe4kcEi8R8:1397195845) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/simulation/network.go#L183)):*
> ```Go
> 	for _, node := range n.Nodes {
> 		if err := n.StartNode(node); err != nil {
> 			log.Fatal(context.TODO(), err)
> ```

can this return an error instead?

---

*[storage/raft_transport.go, line 50 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbkLoCz6tdg8jSpKJ3:-KNbkLoCz6tdg8jSpKJ4:-1841622479) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/storage/raft_transport.go#L50)):*
> ```Go
> )
> 
> var (
> ```

move this to the const block above

---

*[storage/raft_transport.go, line 186 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbktP0hMPdhQdfssPv:-KNbktP0hMPdhQdfssPw:-88077378) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/storage/raft_transport.go#L186)):*
> ```Go
> 
> // This method handles backoff / retry for nodes which aren't being
> // communicated by gossip (yet -- as in the case of reconstituting a
> ```

technically this doesn't know anything about gossip.

---

*[storage/raft_transport_test.go, line 339 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbl5y8oEEyWBund8U7:-KNbl5y8oEEyWBund8U8:-1787544169) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/storage/raft_transport_test.go#L339)):*
> ```Go
> 
> 	for i := 0; i < numMessages; i++ {
> 		sendRaftMessage(t, clientNodeID, nodeID, 1, 2, 1, 2, i, clientTransport)
> ```

this is pretty hard to read now that it's just a bunch of numbers. also, where did these magic numbers come from?

---

*[storage/raft_transport_test.go, line 380 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblJqAyfhdjxjahv7R:-KNblJqAyfhdjxjahv7S:972557073) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/storage/raft_transport_test.go#L380)):*
> ```Go
> 
> 	// Send two messages from client to two different replicas.
> 	sendRaftMessage(t, clientNodeID, nodeID, 1, 2, 1, 2, 0, clientTransport)
> ```

more magic numbers

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[storage/raft_transport.go, line 185 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblkILdJ0qJpfjz28B:-KNblkILdJ0qJpfjz28C:-330799482) ([raw file](https://github.com/cockroachdb/cockroach/blob/01256109e6ad6c7bd8b0a7ef7c06fd7a61607f34/storage/raft_transport.go#L185)):*
> ```Go
> // communicated by gossip (yet -- as in the case of reconstituting a
> // cluster from copied data or after a long downtime)
> func (t *RaftTransport) resolveNodeID(nodeID roachpb.NodeID) (net.Addr, error) {
> ```

 This is a lot of complexity for removing  some log spam. Won't we leak those goroutines if a server goes down and never comes back? Why don't we just put the spammy logging behind a circuit breaker (https://github.com/rubyist/circuitbreaker)?

---

*[storage/raft_transport_test.go, line 362 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbkuM-ZckONDEJ0OPs:-KNbkuM-ZckONDEJ0OPt:2040164355) ([raw file](https://github.com/cockroachdb/cockroach/blob/01256109e6ad6c7bd8b0a7ef7c06fd7a61607f34/storage/raft_transport_test.go#L362)):*
> ```Go
> 	storage.InitialResolveBackoff = 1 * time.Millisecond
> 	defer func() {
> 		storage.InitialResolveBackoff = oldIRB
> ```

`InitialResolveBackoff` should be a property of `*RaftTransport` (or wherever else it fits) and not a global which we manipulate ad-hoc.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 5 of 8 files at r1, 3 of 4 files at r2.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[gossip/gossip.go, line 243 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNcFIeDMqtYFhbC6Net:-KNcFIeDMqtYFhbC6Neu:-2072699240) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/gossip.go#L243)):*
> ```Go
> 
> // SetStallInterval sets the interval between successive checks
> // to determin whether this host is not connected to the gossip
> ```

s/determin/determine/

---

*[gossip/storage_test.go, line 27 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNcGKR4bY1U8RPEb3-J:-KNcGKR4bY1U8RPEb3-K:-1756708947) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/storage_test.go#L27)):*
> ```Go
> 
> 	"github.com/pkg/errors"
> 	"github.com/prometheus/common/log"
> ```

That's not the log package you're looking for.

---

*[storage/raft_transport.go, line 185 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblkILdJ0qJpfjz28B:-KNcIESCYKbl2qf7r90Q:-1499742810) ([raw file](https://github.com/cockroachdb/cockroach/blob/01256109e6ad6c7bd8b0a7ef7c06fd7a61607f34/storage/raft_transport.go#L185)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
>  This is a lot of complexity for removing  some log spam. Won't we leak those goroutines if a server goes down and never comes back? Why don't we just put the spammy logging behind a circuit breaker (https://github.com/rubyist/circuitbreaker)?
</details>
Yeah, we need some limit on the number of retries so we eventually stop retrying dead servers after the upper layers stop trying to send messages to them. 

---

*[storage/raft_transport.go, line 196 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNcIvAlcv7empOst0we:-KNcIvAlcv7empOst0wf:-1636086616) ([raw file](https://github.com/cockroachdb/cockroach/blob/01256109e6ad6c7bd8b0a7ef7c06fd7a61607f34/storage/raft_transport.go#L196)):*
> ```Go
> 		select {
> 		case <-ch:
> 			return t.resolver(nodeID)
> ```

If this attempt fails, it will be logged in a spammy way. If we're going to introduce a retry loop, it should become the only place we call `t.resolver()` and we save its result somewhere for other threads to use.

---

*[storage/raft_transport.go, line 204 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNcK8d8j9OVtswOZH7l:-KNcK8d8j9OVtswOZH7m:1558256101) ([raw file](https://github.com/cockroachdb/cockroach/blob/01256109e6ad6c7bd8b0a7ef7c06fd7a61607f34/storage/raft_transport.go#L204)):*
> ```Go
> 	defer func() {
> 		t.mu.Lock()
> 		delete(t.mu.addrLookups, nodeID)
> ```

Close the channel here instead of where you have it below. This will guarantee that we never orphan an unclosed channel which would leave other goroutines to hang.

---

*[storage/raft_transport_test.go, line 362 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbkuM-ZckONDEJ0OPs:-KNcNOaAAmwba8I2ge0N:138174571) ([raw file](https://github.com/cockroachdb/cockroach/blob/01256109e6ad6c7bd8b0a7ef7c06fd7a61607f34/storage/raft_transport_test.go#L362)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `InitialResolveBackoff` should be a property of `*RaftTransport` (or wherever else it fits) and not a global which we manipulate ad-hoc.
</details>
Yeah, we shouldn't introduce more of these magic globals.

---

*[storage/raft_transport_test.go, line 275 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNcKmOI7ce39Qp-e0mD:-KNcKmOI7ce39Qp-e0mE:-1110269391) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/storage/raft_transport_test.go#L275)):*
> ```Go
> }
> 
> func sendRaftMessage(t *testing.T, fromNID, toNID roachpb.NodeID, fromSID, toSID roachpb.StoreID,
> ```

The order of these parameters is surprising - I'd expect all the "from" ones together, then all the "to" ones. Or even better, pass in a ReplicaDescriptor instead of just its component parts.

---

*[storage/raft_transport_test.go, line 297 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNcL3YrdbY_1DrLKqE0:-KNcL3YrdbY_1DrLKqE1:925485538) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/storage/raft_transport_test.go#L297)):*
> ```Go
> 	if !client.MakeSender(func(err error, _ roachpb.ReplicaDescriptor) {
> 		if err != nil && !grpcutil.IsClosedConnection(err) {
> 			panic(err)
> ```

Why panic when you have a `testing.T` available?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[storage/raft_transport.go, line 185 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblkILdJ0qJpfjz28B:-KNgEsxJZkM0yt_7uZ0Z:-1570946012) ([raw file](https://github.com/cockroachdb/cockroach/blob/01256109e6ad6c7bd8b0a7ef7c06fd7a61607f34/storage/raft_transport.go#L185)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Yeah, we need some limit on the number of retries so we eventually stop retrying dead servers after the upper layers stop trying to send messages to them. 
</details>
What are your thoughts on circuit breaking? I'd be more than happy to just drop messages - Raft can handle it after all, and I feel fairly strongly that the less moving parts, the better.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[storage/raft_transport.go, line 185 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblkILdJ0qJpfjz28B:-KNgW46pQHOdL0dWqKJ6:-1638682916) ([raw file](https://github.com/cockroachdb/cockroach/blob/01256109e6ad6c7bd8b0a7ef7c06fd7a61607f34/storage/raft_transport.go#L185)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What are your thoughts on circuit breaking? I'd be more than happy to just drop messages - Raft can handle it after all, and I feel fairly strongly that the less moving parts, the better.
</details>
I like the idea of circuit breakers, though I've never actually used them. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[storage/raft_transport.go, line 185 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblkILdJ0qJpfjz28B:-KNh0-6JAfl6eUP5Q97W:-485711229) ([raw file](https://github.com/cockroachdb/cockroach/blob/01256109e6ad6c7bd8b0a7ef7c06fd7a61607f34/storage/raft_transport.go#L185)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I like the idea of circuit breakers, though I've never actually used them. 
</details>
Same.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 3 of 8 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[gossip/gossip.go, line 460 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblVWY_L40McLHqImQ:-KNd0WONNKXj-DZtoye2:296882782) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/gossip.go#L460)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: we shouldn't be introducing new `context.TODO()` wherever we're logging anything. can a context be plumbed here?
</details>
That's another change entirely. All of gossip needs to be plumbed.

---

*[gossip/storage_test.go, line 230 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbjdJK-q4hVxtJ2Wjx:-KNd1019O1YXKVVZdlH6:-1137699106) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/storage_test.go#L230)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why `% 3`? this `3` is `len(network.Nodes)`, right? avoid magic numbers.
</details>
Yes. Added a const.

---

*[gossip/storage_test.go, line 250 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbjTS5hf3XGtoi7UIH:-KNd19POeI_QBas9Pkg-:216028858) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/storage_test.go#L250)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why not `for i, p := range stores`?
</details>
No idea; just was copying previous code. Changed.

---

*[gossip/simulation/network.go, line 183 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbkG6BaYxe4kcEi8R7:-KNd1MIiiXbrUMpM-ypk:1223896378) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/simulation/network.go#L183)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can this return an error instead?
</details>
Yes it could, but this is test code. This is another change. Added a TODO.

---

*[storage/raft_transport.go, line 185 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblkILdJ0qJpfjz28B:-KNm9mFd8-3ZxJMR92i_:-61117034) ([raw file](https://github.com/cockroachdb/cockroach/blob/01256109e6ad6c7bd8b0a7ef7c06fd7a61607f34/storage/raft_transport.go#L185)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Same.
</details>
Upon re-inspection, I think we can eliminate the complexity this change originally proposed and instead do the resolve / dial once per remote node. I'm now using a circuit breaker per node ID to avoid busy attempts to establish connections. I'm using the state in the circuit breaker to log failure and success only on state changes, solving the busy logging problem.

---

*[storage/raft_transport.go, line 50 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbkLoCz6tdg8jSpKJ3:-KNd2mblXyIIT6W6l5-4:1815354035) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/storage/raft_transport.go#L50)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> move this to the const block above
</details>
Need to change it in the unittest.

---

*[storage/raft_transport.go, line 186 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbktP0hMPdhQdfssPv:-KNd1_pa_3l5anX_EV4H:1405130756) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/storage/raft_transport.go#L186)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> technically this doesn't know anything about gossip.
</details>
I'm personally not of the opinion that should stop comments from mentioning likely reasons why you'd get backoff here. Otherwise, it's constantly an exercise left to the reader, meaning in practice nobody understands the "why?".

But to your point, I made this less a statement about gossip and more about resolving, though I added the example of gossip in a parenthetical.

---

*[storage/raft_transport_test.go, line 362 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbkuM-ZckONDEJ0OPs:-KNnm1H5txJ11v6H_KZv:189548703) ([raw file](https://github.com/cockroachdb/cockroach/blob/01256109e6ad6c7bd8b0a7ef7c06fd7a61607f34/storage/raft_transport_test.go#L362)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Yeah, we shouldn't introduce more of these magic globals.
</details>
This has been removed.

---

*[storage/raft_transport_test.go, line 297 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNcL3YrdbY_1DrLKqE0:-KNnlr73WipkxmEAHgZS:710164311) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/storage/raft_transport_test.go#L297)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why panic when you have a `testing.T` available?
</details>
I just copied this code. Changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[gossip/gossip.go, line 460 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblVWY_L40McLHqImQ:-KNoCe6NtVq6dcBbaKV7:461511028) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/gossip.go#L460)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> That's another change entirely. All of gossip needs to be plumbed.
</details>
That's true, but we gotta start somewhere, and the caller of this method already has a context nearby (for its own logging).

---

*[gossip/storage_test.go, line 250 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbjTS5hf3XGtoi7UIH:-KNoCx2JeupzhCac1bMc:-296663643) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/storage_test.go#L250)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> No idea; just was copying previous code. Changed.
</details>
oh, maybe testStorage is not safe to copy.

---

*[gossip/simulation/network.go, line 183 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbkG6BaYxe4kcEi8R7:-KNoDGfhjgkmmW1a8PMh:-1848964596) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/simulation/network.go#L183)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Yes it could, but this is test code. This is another change. Added a TODO.
</details>
I don't get it - this is new code in this PR; how is making it return an error another change?

---

*[storage/raft_transport.go, line 321 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNoEJxaeCEy9jp6fD4Y:-KNoEJxaeCEy9jp6fD4Z:1554031884) ([raw file](https://github.com/cockroachdb/cockroach/blob/a18fb5da0430fdd9eeaf0865fe56c986d7e7253f/storage/raft_transport.go#L321)):*
> ```Go
> 	isSnap := req.Message.Type == raftpb.MsgSnap
> 	toReplica := req.ToReplica
> 	// Get a connection to the node specified by the replica's node
> ```

is this right? if we have room in the queue we should probably not drop the message, even if it might not go out for a bit.

---

*[storage/raft_transport_test.go, line 380 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblJqAyfhdjxjahv7R:-KNoDU67Msag3OqHrZUx:1544562890) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/storage/raft_transport_test.go#L380)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> more magic numbers
</details>
I see that they're named now, but where did these numbers come from?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[gossip/gossip.go, line 460 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblVWY_L40McLHqImQ:-KNqpWsWhwmsvVNo_5ya:84497332) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/gossip.go#L460)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> That's true, but we gotta start somewhere, and the caller of this method already has a context nearby (for its own logging).
</details>
Where is the nearby context? There don't seem to be any in gossip.

---

*[gossip/storage_test.go, line 250 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbjTS5hf3XGtoi7UIH:-KNqppz-lMRCOclIc25r:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/storage_test.go#L250)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> oh, maybe testStorage is not safe to copy.
</details>
Done.

---

*[gossip/simulation/network.go, line 183 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNbkG6BaYxe4kcEi8R7:-KNqprJQ2y8Fvbpu8Zjs:-27572864) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/simulation/network.go#L183)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I don't get it - this is new code in this PR; how is making it return an error another change?
</details>
My point is that the startup of the simulated network doesn't return an error (`SimulateNetwork`, which is called from various places). This code is really old and needs some updating, but I don't want to just let it creep into this PR. The change here merely moves code around so that `StartNode()` isn't invoked automatically from `NewNetwork()`.

---

*[storage/raft_transport.go, line 321 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNoEJxaeCEy9jp6fD4Y:-KNqqavIhviKtbN9D7Fv:-124718865) ([raw file](https://github.com/cockroachdb/cockroach/blob/a18fb5da0430fdd9eeaf0865fe56c986d7e7253f/storage/raft_transport.go#L321)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this right? if we have room in the queue we should probably not drop the message, even if it might not go out for a bit.
</details>
This was @tschottdorf's suggestion ("What are your thoughts on circuit breaking? I'd be more than happy to just drop messages - Raft can handle it after all, and I feel fairly strongly that the less moving parts, the better."). I think this is not just a good idea, it's actually a requirement. To see why, consider:

The queue of outgoing Raft messages helps prevent unnecessary drops if a remote node is slow on the uptake. However, keep in mind that there is one queue _per_ outgoing replica. Each queue can hold up to 100 messages. If a node goes AWOL, that means that every single replica held on that node will fill up to 100 messages. Not good.

---

*[storage/raft_transport_test.go, line 380 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblJqAyfhdjxjahv7R:-KNqruqi0LI4uLMUxD77:-1410066377) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/storage/raft_transport_test.go#L380)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I see that they're named now, but where did these numbers come from?
</details>
There's unused as there is no corresponding replica; these messages are never passed through to a store; they're just placeholders so the raft message isn't rejected. We're only checking whether the raft message is sent and delivered.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[gossip/gossip.go, line 460 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNblVWY_L40McLHqImQ:-KNqtUgDIkmsv_VGWdWL:2120954902) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cfcc739c59e8db2dd7f77ff86b759e8259a41f2/gossip/gossip.go#L460)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Where is the nearby context? There don't seem to be any in gossip.
</details>
Left a comment on it.

---

*[gossip/gossip.go, line 967 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNqt7MVmgDKIOYSO0YZ:-KNqt7MWIM8kmcDMFwLV:1646191888) ([raw file](https://github.com/cockroachdb/cockroach/blob/a18fb5da0430fdd9eeaf0865fe56c986d7e7253f/gossip/gossip.go#L967)):*
> ```Go
> 		g.maybeCleanupBootstrapAddresses()
> 		if g.stalled {
> 			log.Infof(context.TODO(), "node has connected to cluster via gossip")
> ```

@spencerkimball the context I'm referring to is here. You need only extract a local and pass it to `maybeCleanupBoostrapAddresses`. Plumbing further up is not required.

---

*[storage/raft_transport.go, line 321 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNoEJxaeCEy9jp6fD4Y:-KNqtItJr6XRaWJ24yjz:-1538074712) ([raw file](https://github.com/cockroachdb/cockroach/blob/a18fb5da0430fdd9eeaf0865fe56c986d7e7253f/storage/raft_transport.go#L321)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> This was @tschottdorf's suggestion ("What are your thoughts on circuit breaking? I'd be more than happy to just drop messages - Raft can handle it after all, and I feel fairly strongly that the less moving parts, the better."). I think this is not just a good idea, it's actually a requirement. To see why, consider:
> 
> The queue of outgoing Raft messages helps prevent unnecessary drops if a remote node is slow on the uptake. However, keep in mind that there is one queue _per_ outgoing replica. Each queue can hold up to 100 messages. If a node goes AWOL, that means that every single replica held on that node will fill up to 100 messages. Not good.
</details>
Seem to me that there are cases where many of these outgoing queues can fill up even in the absence of a full disconnect, so I'm not convinced about this "solution", but OK.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 7 of 9 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[gossip/gossip.go, line 967 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNqt7MVmgDKIOYSO0YZ:-KNqw9RC_ThL4xuUOrsu:-1872512853) ([raw file](https://github.com/cockroachdb/cockroach/blob/a18fb5da0430fdd9eeaf0865fe56c986d7e7253f/gossip/gossip.go#L967)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> @spencerkimball the context I'm referring to is here. You need only extract a local and pass it to `maybeCleanupBoostrapAddresses`. Plumbing further up is not required.
</details>
You mean create a local context here via `context.TODO()` and then pass that?

Btw, I had some ideas about how to make good use of long-running contexts for things like gossip which aren't necessarily part of a short-lived request. If you come in today, let's discuss.

---

*[storage/raft_transport.go, line 321 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNoEJxaeCEy9jp6fD4Y:-KNqwV3pQXeJF0CC3VYK:-1728504050) ([raw file](https://github.com/cockroachdb/cockroach/blob/a18fb5da0430fdd9eeaf0865fe56c986d7e7253f/storage/raft_transport.go#L321)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Seem to me that there are cases where many of these outgoing queues can fill up even in the absence of a full disconnect, so I'm not convinced about this "solution", but OK.
</details>
I don't doubt there are other sticky situations a node might get into which would fill up all of the replica queues destined for it. However, one thing at a time. This change is meant to stop spammy logging, and it's also true that a node simply going away is by far the most common situation and this change also prevents that from causing replica queues from filling up.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[gossip/gossip.go, line 967 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNqt7MVmgDKIOYSO0YZ:-KNqx99k78kSzvtiZKM-:-1155450458) ([raw file](https://github.com/cockroachdb/cockroach/blob/a18fb5da0430fdd9eeaf0865fe56c986d7e7253f/gossip/gossip.go#L967)):*
> You mean create a local context here via context.TODO() and then pass that?

Yep, exactly.

> Btw, I had some ideas about how to make good use of long-running contexts for things like gossip which aren't necessarily part of a short-lived request. If you come in today, let's discuss.

Sounds good, will be in this afternoon.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[gossip/simulation/network.go, line 47 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNrrsF1yUb_ELO3auR9:-KNrrsF1yUb_ELO3auRA:1337093386) ([raw file](https://github.com/cockroachdb/cockroach/blob/60ba3ef166bd97991c5654ba599efabc08c9ed93/gossip/simulation/network.go#L47)):*
> ```Go
> 	Gossip   *gossip.Gossip
> 	Server   *grpc.Server
> 	Addr     net.Addr
> ```

this is now redundant

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-:-KNsEM5G3JLdrPBturO7:1417088721)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[GLOCKFILE, line 92 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNsJznMFOIxLqTDN1vR:-KNsJznMFOIxLqTDN1vS:1180302740) ([raw file](https://github.com/cockroachdb/cockroach/blob/642365dce2bfc78a1c53962c371e00165b19b2b6/GLOCKFILE#L92)):*
> ```
> github.com/rcrowley/go-metrics bdb33529eca3e55eac7328e07c57012a797af602
> github.com/robfig/glock c7fb89f56fbead624f7424f81f70af3a742433c5
> github.com/rubyist/circuitbreaker 434e9dfaa002281310291f4042827d0940526785
> ```

I noticed we have our own fork of circuitbreaker, should we use that?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[GLOCKFILE, line 54 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNsLmzl7n5BFMagjM5G:-KNsLmzmO2-SekdQS_si:1243266820) ([raw file](https://github.com/cockroachdb/cockroach/blob/642365dce2bfc78a1c53962c371e00165b19b2b6/GLOCKFILE#L54)):*
> ```
> github.com/elastic/gosigar 1d5bcdfbb4e5b8b5982f1e87ee6f13a806dd9aa0
> github.com/elazarl/go-bindata-assetfs 57eb5e1fc594ad4b0b1dbea7b286d299e0cb43c2
> github.com/facebookgo/clock 600d898af40aa09a7a93ecb9265d87b0504b6f03
> ```

Do we still need this one and the g2s? I didn't find any references in the latest version

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[GLOCKFILE, line 54 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNsLmzl7n5BFMagjM5G:-KNsMiV6YZpkYLglSvxd:-642049031) ([raw file](https://github.com/cockroachdb/cockroach/blob/642365dce2bfc78a1c53962c371e00165b19b2b6/GLOCKFILE#L54)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Do we still need this one and the g2s? I didn't find any references in the latest version
</details>
Ah `facebookgo/clock` is needed by circuitbreaker. I didn't find any indirect use of `g2s` though I might have missed it

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[GLOCKFILE, line 92 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNsJznMFOIxLqTDN1vR:-KNsOR-s7Q1ij1rZzIPO:-785150401) ([raw file](https://github.com/cockroachdb/cockroach/blob/642365dce2bfc78a1c53962c371e00165b19b2b6/GLOCKFILE#L92)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> I noticed we have our own fork of circuitbreaker, should we use that?
</details>
Looks like spencer forked to send https://github.com/rubyist/circuitbreaker/pull/38 which is already merged.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[GLOCKFILE, line 54 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNsLmzl7n5BFMagjM5G:-KNsZpVU82jzczPQDcfY:-88209735) ([raw file](https://github.com/cockroachdb/cockroach/blob/642365dce2bfc78a1c53962c371e00165b19b2b6/GLOCKFILE#L54)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Ah `facebookgo/clock` is needed by circuitbreaker. I didn't find any indirect use of `g2s` though I might have missed it
</details>
Ah, it's used by `circuitbreaker`'s tests. I think we can remove it.

---

*[GLOCKFILE, line 92 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNsJznMFOIxLqTDN1vR:-KNsZT6q6goHlKyuKZCV:-895004954) ([raw file](https://github.com/cockroachdb/cockroach/blob/642365dce2bfc78a1c53962c371e00165b19b2b6/GLOCKFILE#L92)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Looks like spencer forked to send https://github.com/rubyist/circuitbreaker/pull/38 which is already merged.
</details>
Yeah, that fork should be retired.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021)*
<!-- Sent from Reviewable.io -->

tamird commented:
let's not start mucking with GLOCKFILE by hand.

On Fri, Jul 29, 2016 at 5:33 PM, Spencer Kimball <notifications@github.com>
wrote:

> Review status: all files reviewed at latest revision, 10 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *GLOCKFILE, line 54 [r7]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNsLmzl7n5BFMagjM5G:-KNsZpVU82jzczPQDcfY:-88209735>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/642365dce2bfc78a1c53962c371e00165b19b2b6/GLOCKFILE#L54>):*
> *Previously, RaduBerinde wrote…*
>
> Ah facebookgo/clock is needed by circuitbreaker. I didn't find any
> indirect use of g2s though I might have missed it
>
>
> Ah, it's used by circuitbreaker's tests. I think we can remove it.
>
> ------------------------------
>
> *GLOCKFILE, line 92 [r7]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNsJznMFOIxLqTDN1vR:-KNsZT6q6goHlKyuKZCV:-895004954>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/642365dce2bfc78a1c53962c371e00165b19b2b6/GLOCKFILE#L92>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> Looks like spencer forked to send rubyist/circuitbreaker#38
> <https://github.com/rubyist/circuitbreaker/pull/38> which is already
> merged.
>
>
> Yeah, that fork should be retired.
>
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/8021>*
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/8021#issuecomment-236299239>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPOQghCWefnpNOhgsXgVD8ZIYSbf9ks5qanG1gaJpZM4JUtgP>
> .
>

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 5 files at r3, 1 of 5 files at r5, 4 of 4 files at r7, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[storage/raft_transport.go, line 199 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-KNszRI3mL2KjdP13sLO:-KNszRI3mL2KjdP13sLP:-250791877) ([raw file](https://github.com/cockroachdb/cockroach/blob/591d40743670fa6ecd5c973f22cfd03fd5d6f24e/storage/raft_transport.go#L199)):*
> ```Go
> 		// intervals between successive attempts to resolve a node address
> 		// and connect via GRPC.
> 		expBO := backoff.NewExponentialBackOff()
> ```

I think we should set all the major parameters here instead of using the defaults for some of them; it's easier than flipping back to the `backoff` source and protects us against any upstream changes to the default value.

500ms-1s is a narrow range for an exponential backoff; I think a smaller minimum and higher maximum would make sense (or else a simpler constant backoff)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8021#-:-KNt-WLgF4TfmrTRIwbA:1417088721)*
<!-- Sent from Reviewable.io -->
